File: 백준/Silver/24523. 내 뒤에 나와 다른 수/내 뒤에 나와 다른 수.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static int n;
    static int arr[] = new int[1000001];

    public static void solve() {
        int ans[] = new int[1000001];
        int mem = -1;
        for (int i = n; i >= 1; i--) {
            ans[i] = mem;
            if (arr[i] != arr[i - 1]) {
                mem = i;
            }
        }
        StringBuilder sb = new StringBuilder();
        for (int i = 1; i <= n; i++) {
            sb.append(ans[i]).append(" ");
        }
        System.out.println(sb);

    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());

        for (int i = 1; i <= n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        solve();
    }
}

File: 백준/Silver/24523. 내 뒤에 나와 다른 수/내 뒤에 나와 다른 수.cc
#include <bits/stdc++.h>
using namespace std;

int n;
int arr[1000001], ans[1000001];

void solve() {
  int mem = -1;
  for (int i = n; i >= 1; i--) {
    ans[i] = mem;
    if (arr[i - 1] != arr[i]) mem = i;
  }

  for (int i = 1; i <= n; i++) {
    cout << ans[i] << " ";
  }
  cout << "\n";
}

int main() {
  ios::sync_with_stdio(false), cin.tie(0), cout.tie(0); /* FastIO */

  cin >> n;
  for (int i = 1; i <= n; i++) {
    cin >> arr[i];
  }

  solve();
}
